### PR TITLE
Feature/sbd chapter

### PR DESCRIPTION
**Pull request by taroth21**
@gao-yan, @tomschr : please review! Thanks in advance!

Summary of Changes: Add missing information and corrections by ygao (as discusssed on ML/IRC/PR#72 today)
-  diskless SBD needs no fencing resource
-> move config of fencing rsc and cluster props to dedicated section about SBD with 1-3 devices
-> add required steps about cluster props to procedure for diskless SBD
- mention more details about stonith-watchdog-timeout:
  * for SBD with devices (set to zero)
  * for diskless SBD: >= SBD_WATCHDOG_TIMEOUT (or set to negative value for auto-calculation, Fate#324508)
- move commented information from abstract to diskless SBD section for better understanding plus add some information by ygao
- add remarks to testing section (remaining comments  from PR#72 that have not been solved yet)

DRAFT PDF version attached [cha.ha.storage.protect_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/1947419/cha.ha.storage.protect_remarks_color_draft_en.pdf)

### CODE REVIEW

File: xml/ha_storage_protection.xml [hunk 1]
```diff
@@ -842,21 +793,128 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+    node needs to be fenced.</para>
+  </sect1>
+
+  <sect1 xml:id="sec.ha.storage.protect.diskless-sbd">
+   <title>Setting Up Diskless SBD</title>
+   <para>SBD can be operated in a diskless mode. In this mode, a watchdog device
+    will be used to reset the node in the following cases: if it loses quorum,
+    if any monitored daemon is lost and not recovered or if Pacemaker decides
+    that the node requires fencing. Diskless SBD is basically based on
+     <quote>self-fencing</quote> of a node, depending on the status of the quorum
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: status of cluster, quorum and ...

File: xml/ha_storage_protection.xml [hunk 2]
```diff
@@ -842,21 +793,128 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+SBD_WATCHDOG_DEV=/dev/watchdog
+SBD_WATCHDOG_TIMEOUT=5</screen>
+      <para>
+       The <varname>SBD_DEVICE</varname> entry is not needed as no shared
+       disk is used. When this parameter is missing, the <systemitem>sbd</systemitem>
+       service does not start any watcher process for SBD devices.
+      </para>
+    </step>
+    <step>
+     <para>On each node, enable the SBD service to start at boot time: </para>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: enable the SBD service to start when starting cluster stack

-- sbd.service has this special configuration "PartOf=corosync.service". So enabling sbd.service doesn't necessarily means it'll start at boot time. It only means it'll start together with corosync.service. They are actually brought up by pacemaker.service by dependencies.

File: xml/ha_storage_protection.xml [hunk 3]
```diff
@@ -879,12 +937,18 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
       about this step. Taken from FATE entry, developers could add more info.
      </remark>
      <para>
+      <remark>taroth 2018-04-28: I would list this as steps:
+       1. Configure a <property>on-fail=fence</property> property for a resource monitor operation: [...]
+          (not sure if giving the line is enough, perhaps also mention command with which to set this property?)
+       2. Produce a failure of the monitoring operation (how???).
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: Hrm, quite depends on the resource.  Usually by killing the daemon of it's a service.

File: xml/ha_storage_protection.xml [hunk 4]
```diff
@@ -348,6 +343,16 @@
        </para>
       </listitem>
      </varlistentry>
+     <varlistentry>
+      <term><literal>stonith-watchdog-timeout</literal> in the CIB</term>
+      <listitem>
+       <para>
+        This timeout is set in the CIB as global cluster property. If not set
+        explicitly, it defaults to <literal>0</literal>, which is appropriate for
+        using SBD with 1-3 devices. For use of SBD in diskless mode, see FIXME
```
Comment (tomschr) [inline xml/ha_storage_protection.xml]: That FIXME shouldn't be there I guess. :wink:

File: xml/ha_storage_protection.xml [hunk 5]
```diff
@@ -348,6 +343,16 @@
...
+        for more details.</para>
+      </listitem>
+     </varlistentry>
     </variablelist>
   <para>
    If you change the watchdog timeout, you need to adjust the other two timeouts
```
Comment (tomschr) [inline xml/ha_storage_protection.xml]: Should we mention somewhere the relationship between positive timeouts and negative timeouts?

File: xml/ha_storage_protection.xml [hunk 6]
```diff
@@ -842,21 +793,128 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+       the fencing target should be fenced and the online nodes consider the
+       fencing target has been fenced after the time stonith-watchdog-timeout.</remark></para>
     </listitem>
     <listitem>
-     <para>Simulate a SBD failure on the SBD inquisitor and use its process ID:</para>
+     <para>Simulate a SBD failure on the SBD inquisitor and use its process ID:
+      <remark>taroth 2018-04-28: struggling a little bit with the order of this...
+      How about the following? "Simulate an SBD failure by killing the SBD
+      inquisitor process: 1. Identify the process ID of the SBD inquisitor with
+      the following command: [...] 2. Kill the process: [...] Also here, would
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: Sounds good.

File: xml/ha_storage_protection.xml [hunk 7]
```diff
@@ -842,21 +793,128 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
     configured for use of SBD and will use this method in case a
-    node needs to be fenced. </para>
+    node needs to be fenced.</para>
+  </sect1>
+
+  <sect1 xml:id="sec.ha.storage.protect.diskless-sbd">
+   <title>Setting Up Diskless SBD</title>
+   <para>SBD can be operated in a diskless mode. In this mode, a watchdog device
+    will be used to reset the node in the following cases: if it loses quorum,
+    if any monitored daemon is lost and not recovered or if Pacemaker decides
```
Comment (tomschr) [inline xml/ha_storage_protection.xml]: "... if it loses quorum, if and monitored daemon..." sounds weird to me. "or" missing?

File: xml/ha_storage_protection.xml [hunk 8]
```diff
@@ -842,21 +793,128 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+     </callout>
+     <callout arearefs="co.ha.sbd.watchdog-timeout">
+      <para>For diskless SBD, this parameter must not equal zero.
+       It defines after how long it is assumed that the fencing target has already
+       self-fenced. Therefore its value needs to be &gt;= the value of
+       <varname>SBD_WATCHDOG_TIMEOUT</varname> in <filename>/etc/sysconfig/sbd</filename>.
+       Starting with &productname; 15, if you set it to a negative value, Pacemaker
+       will automatically calculate an appropriate timeout based on the value of
+       <parameter>SBD_WATCHDOG_TIMEOUT</parameter> that is configured in
+       <filename>/etc/sysconfig/sbd</filename>.
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: Or to be precise, twice of SBD_WATCHDOG_TIMEOUT as also described in pacemaker/doc/Pacemaker_Explained/en-US/Ch-Options.txt:
https://github.com/ClusterLabs/pacemaker/commit/cbf53fe834377476d9edd713d4bd94998d938ee4

File: xml/ha_storage_protection.xml (hunk 1 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Done.

File: xml/ha_storage_protection.xml (hunk 2 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Changed in all places in the chapter.

File: xml/ha_storage_protection.xml (hunk 3 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Thanks, added this information.

File: xml/ha_storage_protection.xml (hunk 6 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Done.

File: xml/ha_storage_protection.xml (hunk 8 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Great, added this information.

File: xml/ha_storage_protection.xml (hunk 4 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: Fixed (replaced by xref).

File: xml/ha_storage_protection.xml (hunk 5 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: This is mentioned in the callouts in pro.ha.storage.protect.confdiskless.

File: xml/ha_storage_protection.xml (hunk 7 above)
Comment (taroth21) [reply xml/ha_storage_protection.xml]: No, the 'or' was there (a bit later in the sentence), added a comma to make it clearer.

**Comment by taroth21** [comment]
Latest version (after integrating your corrections, many thanks!):

[cha.ha.storage.protect_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/1950486/cha.ha.storage.protect_remarks_color_draft_en.pdf)

**Comment by gao-yan** [comment]
Awesome work. Thanks!

File: xml/ha_storage_protection.xml [hunk 9]
```diff
@@ -842,22 +800,126 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+      <screen>&prompt.root;<command>crm</command> configure show | grep have-watchdog
+         have-watchdog=true</screen>
+    </step>
+    <step>
+     <para>Run <command>crm configure</command> and set the following cluster
+      properties on the &crmshell;:</para>
+<screen>&prompt.crm.conf;<command>property</command> stonith-enabled="true" <xref
+        linkend="co.ha.sbd.st.enabled" xrefstyle="select:label"/>
+&prompt.crm.conf;<command>property</command> stonith-watchdog-timeout=10 <co
+        xml:id="co.ha.sbd.diskless.watchdog-timeout"/></screen>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: Somehow in the PDF, this is shown to have the conflicting indexing number (1)  which belongs to the previous one.

File: xml/ha_storage_protection.xml [hunk 10]
```diff
@@ -842,22 +800,126 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+    <calloutlist>
+     <callout arearefs="co.ha.sbd.st.enabled">
+      <para>
+       This is the default configuration, because clusters with &stonith; are not supported.
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: s/with/without/

File: xml/ha_storage_protection.xml [hunk 11]
```diff
@@ -842,22 +800,126 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+       But in case &stonith; had been deactivated for testing purposes,
+       make sure this parameter is set to <literal>true</literal>.</para>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: The same again, "again"? :)

File: xml/ha_storage_protection.xml [hunk 12]
```diff
@@ -842,22 +800,126 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+     </callout>
+     <callout arearefs="co.ha.sbd.diskless.watchdog-timeout">
+      <para>For diskless SBD, this parameter must not equal zero.
+       It defines after how long it is assumed that the fencing target has already
+       self-fenced. Therefore its value needs to be &gt;= the value of
+       <varname>SBD_WATCHDOG_TIMEOUT</varname> in <filename>/etc/sysconfig/sbd</filename>.
+       Starting with &productname; 15, if you set it to a negative value, Pacemaker
+       will automatically calculate an appropriate timeout. In that case, it will
+       be set to twice the value of <parameter>SBD_WATCHDOG_TIMEOUT</parameter>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: it will be twice... -- since it won't be explicitly "set" ;)

File: xml/ha_storage_protection.xml [hunk 13]
```diff
@@ -869,25 +931,38 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
-     </para>
-     <screen>&prompt.root;<command>kill</command> -9 1859 </screen>
+      </step>
+      <step>
+       <para>Simulate an SBD failure by killing the SBD inquisitor process.
+       In our example, the process ID of the SBD inquisitor is
+         <literal>1859</literal>):</para>
+       <screen>&prompt.root;<command>kill</command> -9 1859 </screen>
+       <para>
+        <remark>taroth 2018-04-28: @DEVs, What is the expected result?</remark>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: The node proactively self-fences. And the other nodes notices the loss of the node and consider it has self-fenced after the stonith-watchdog-timeout.

File: xml/ha_storage_protection.xml [hunk 14]
```diff
@@ -869,25 +931,38 @@ SBD_WATCHDOG_TIMEOUT=5</screen>
...
+       </para>
+      </step>
+     </procedure>
     </listitem>
+   </varlistentry>
+   <varlistentry>
+    <term>Producing a Failure of a Resource Operation</term>
```
Comment (gao-yan) [inline xml/ha_storage_protection.xml]: How about "Producing a Failure of a Resource Operation that triggers fencing"

And I think it makes sense to  first mention the *typical* scenario where resource failure triggers fencing, for example:
"With a normal configuration, a failure of a resource stop operation triggers fencing. So you can somehow produce such a resource stop failure, if possible, to trigger fencing. If it's not easy to do so, another option is to temporarily change the configuration of a resource monitor operation and produce a monitor failure like the following to test that."

Comment (taroth21) [reply xml/ha_storage_protection.xml]: Thanks, good idea. Integrated with https://github.com/SUSE/doc-sleha/commit/16c61998b.

**Pull request merged**